wolfcrypt/src/signature.c

```diff
@@ -139,117 +257,36 @@ int wc_SignatureVerifyHash(
-                ret = SIG_VERIFY_E;
+            /* Validate signature len (1 to max is okay) */
+            if ((int)sig_len > wc_SignatureGetSize(sig_type, key, key_len)) {
+                WOLFSSL_MSG("wc_SignatureVerify: Invalid sig type/len");
+                return BAD_FUNC_ARG;
             }
-#else
-            ret = SIG_TYPE_E;
-#endif
+            ret = _SignatureVerifyECC(hash_data, hash_len, (byte*)sig, sig_len,
+                    key);
             break;
-        }
 
         case WC_SIGNATURE_TYPE_RSA_W_ENC:
         case WC_SIGNATURE_TYPE_RSA:
-        {
-#ifndef NO_RSA
-    #ifdef WOLFSSL_CRYPTOCELL
-        if (sig_type == WC_SIGNATURE_TYPE_RSA_W_ENC) {
-            ret = cc310_RsaSSL_Verify(hash_data, hash_len, (byte*)sig,
-                (RsaKey*)key, cc310_hashModeRSA(hash_type, 0));
-        }
-        else {
-            ret = cc310_RsaSSL_Verify(hash_data, hash_len, (byte*)sig,
-                (RsaKey*)key, cc310_hashModeRSA(hash_type, 1));
-        }
-    #else
-
-            word32 plain_len = hash_len;
-        #if defined(WOLFSSL_SMALL_STACK) && !defined(WOLFSSL_NO_MALLOC)
-            byte *plain_data;
-        #else
-            byte  plain_data[MAX_ENCODED_SIG_SZ];
-        #endif
-
-            /* Make sure the plain text output is at least key size */
-            if (plain_len < sig_len) {
-                plain_len = sig_len;
-            }
-        #if defined(WOLFSSL_SMALL_STACK) && !defined(WOLFSSL_NO_MALLOC)
-            plain_data = (byte*)XMALLOC(plain_len, NULL, DYNAMIC_TYPE_TMP_BUFFER);
-            if (plain_data)
-        #else
-            if (plain_len <= sizeof(plain_data))
-        #endif
-            {
-                byte* plain_ptr = NULL;
-                XMEMSET(plain_data, 0, plain_len);
-                XMEMCPY(plain_data, sig, sig_len);
-                /* Perform verification of signature using provided RSA key */
-                do {
-                #ifdef WOLFSSL_ASYNC_CRYPT
-                    ret = wc_AsyncWait(ret, &((RsaKey*)key)->asyncDev,
-                        WC_ASYNC_FLAG_CALL_AGAIN);
-                #endif
-                if (ret >= 0)
-                        ret = wc_RsaSSL_VerifyInline(plain_data, sig_len, &plain_ptr, (RsaKey*)key);
-                } while (ret == WC_PENDING_E);
-                if (ret >= 0 && plain_ptr) {
-                    if ((word32)ret == hash_len &&
-                            XMEMCMP(plain_ptr, hash_data, hash_len) == 0) {
-                        ret = 0; /* Success */
-                    }
-                    else {
-                        ret = SIG_VERIFY_E;
-                    }
-                }
-            #if defined(WOLFSSL_SMALL_STACK) && !defined(WOLFSSL_NO_MALLOC)
-                XFREE(plain_data, NULL, DYNAMIC_TYPE_TMP_BUFFER);
-            #endif
-            }
-            else {
-                ret = MEMORY_E;
+            /* RSA signature should be exactly as long as modulus */
```

Comment:
An RSA signature is encoded hash of message raised to the power of the private key modulo n.
That is the result is a number that can have leading zeros that may have been dropped.